pkgs/development/cuda-modules/db/json.nix

```diff
@@ -0,0 +1,276 @@
+  known.systemsNv = [
+    "source"
+    "linux-aarch64"
+    "linux-all"
+    "linux-ppc64le"
+    "linux-sbsa"
+    "linux-x86_64"
+    "windows-x86_64"
+  ];
+  known.badProducts = [
+    "cudnn-v8-9-cuda-12"
+    "cudnn-v8-9-cuda-11"
+    null
+  ];
+  known.products = [
+    "cuda"
+    "cudnn"
+    "cutensor"
+  ];
+  known.toplevel = [
+    "release_date"
+    "release_label"
+    "release_product"
+  ];
+  looksLikePackage =
+    name: p:
+    !(builtins.elem name known.toplevel)
+    && builtins.isAttrs p
+    && builtins.any looksLikeSystem (builtins.attrNames p);
+  looksLikeSystem = lib.flip builtins.elem known.systemsNv;
+  licenseOf =
+    pname: perSystem:
+    lib.addErrorContext "While inferring the license of ${pname}" (
+      if perSystem.license or null != null then
+        perSystem.license
+      else if productName == "cuda" then
+        lib.licenses.nvidiaCudaRedist.shortName
+      else if productName != null then
+        productName
+      else
+        null
+    );
+  # Handle inconsistencies like "Imex" v. "Nvidia-Imex"
+  mkChooseLonger' =
+    maxPrio: s:
+    lib.mkOverride (lib.modules.defaultOverridePriority - (lib.min maxPrio (lib.stringLength s))) s;
+  mkChooseLonger = mkChooseLonger' 40;
+  evaluated =
+    let
+      manifestDecl = ../modules/generic/manifests/redistrib/manifest.nix;
+      featureDecl = ../modules/generic/manifests/feature/manifest.nix;
+    in
+    lib.evalModules {
+      modules = [
+        (builtins.fromJSON (builtins.readFile path) // { _file = path; })
+        {
+          freeformType =
+            if lib.hasPrefix "feature_" (builtins.baseNameOf path) then
+              (types.submodule {
+                imports = [ featureDecl ];
+              })
+            else
+              (types.submodule {
+                imports = [ manifestDecl ];
+              });
+          _file = ./json.nix;
+        }
+      ];
+    };
+  productName =
+    if builtins.elem raw.release_product or null known.badProducts then
+      builtins.head (
+        lib.optionals (raw.release_product or null != null)
+          # Cf. cudnn circa 8.9 (e.g. cudnn-v8-9-cuda-12)
+          (lib.splitString "-" raw.release_product)
+        ++ concatMapRaw (
+          { name, value }:
+          let
+            rawPackage = value;
+          in
+          lib.optionals (looksLikePackage name rawPackage) (
+            lib.optionals (builtins.elem name known.products) [
+            ]
+          )
+        )
+        ++ [ null ]
+      )
+    else
+      raw.release_product;
+  releaseLabel =
+    raw.release_label or (builtins.head (
+      lib.match "^.*_([0-9]{1,2}(\\.[0-9]{1,3}){1,3})\\.json$" (builtins.baseNameOf path) # Forgive me Lord
+    ));
+  distribution_path = if productName == null then null else "${productName}/redist/";
+
+  raw = evaluated.config;
+  concatMapRaw = lib.flip lib.concatMap (lib.attrsToList raw);
+
+  rawPackages = lib.filterAttrs looksLikePackage raw;
+
+  fmapPackages =
```

Comment:
Can you add a comment about what this does (recursive flattening and application of a function) and something about what the expected arity/type of `f` is?